delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/BlazegraphQuery.scala

```diff
@@ -0,0 +1,127 @@
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.CompositeViewProjection.SparqlProjection
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.CompositeViewRejection.{AuthorizationFailed, ProjectionNotFound, WrappedBlazegraphClientError}
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.ProjectionType.SparqlProjectionType
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.{CompositeView, CompositeViewRejection}
+import ch.epfl.bluebrain.nexus.delta.sdk.Acls
+import ch.epfl.bluebrain.nexus.delta.sdk.model.IdSegment
+import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sourcing.config.ExternalIndexingConfig
+import monix.bio.IO
+
+trait BlazegraphQuery {
+
+  /**
+    * Queries the blazegraph common namespace of the passed composite view.
+    * We check for the caller to have the necessary query permissions on all the views' projections before performing the query.
+    *
+    * @param id         the id of the composite view either in Iri or aliased form
+    * @param project    the project where the view exists
+    * @param query      the sparql query to run
+    */
+  def query(
+      id: IdSegment,
+      project: ProjectRef,
+      query: SparqlQuery
+  )(implicit caller: Caller): IO[CompositeViewRejection, SparqlResults]
+
+  /**
+    * Queries the blazegraph namespace of the passed composite views' projection.
+    * We check for the caller to have the necessary query permissions on the views' projections before performing the query.
+    *
+    * @param id           the id of the composite view either in Iri or aliased form
+    * @param projectionId the id of the composite views' target projection either in Iri or aliased form
+    * @param project      the project where the view exists
+    * @param query        the sparql query to run
+    */
+  def query(
+      id: IdSegment,
+      projectionId: IdSegment,
+      project: ProjectRef,
+      query: SparqlQuery
+  )(implicit caller: Caller): IO[CompositeViewRejection, SparqlResults]
+
+  /**
+    * Queries all the blazegraph namespaces of the passed composite views' projection.
+    * We check for the caller to have the necessary query permissions on the views' projections before performing the query.
+    *
+    * @param id           the id of the composite view either in Iri or aliased form
+    * @param project      the project where the view exists
+    * @param query        the sparql query to run
+    */
+  def queryProjections(
+      id: IdSegment,
+      project: ProjectRef,
+      query: SparqlQuery
+  )(implicit caller: Caller): IO[CompositeViewRejection, SparqlResults]
+
+}
+
+object BlazegraphQuery {
+
+  final def apply(
+      acls: Acls,
+      views: CompositeViews,
+      client: BlazegraphClient
```

Comment:
The goal here is more about filtering projections before querying than query than querying BG itself.

If we only pass a function (namespaces, query) => IO[CompositeViewRejection, SparqlResults], things are gonna be easier in tests as we won't have to spin up docker and a client and everything.

Even to fetch the view, it could be a function

Reply:
Yes I know. However I'm keeping it this way, since it is how it is done in `ElasticSearchViewsQuery` and `BlazegraphViewsQuery`

Reply:
It would indeed make things simpler...

Reply:
Yes. I'm working on that